jwst/badpix_selfcal/badpix_selfcal_step.py

```diff
@@ -0,0 +1,210 @@
+    force_single = boolean(default=False)  #force single input exposure
+    save_flagged_bkg = boolean(default=False)  #save flagged background exposures to file
+    skip = boolean(default=True)
+    """
+
+    def save_model(self, model, *args, **kwargs):
+        """Override save_model to suppress index 0 when save_model is True
+        """
+        kwargs["idx"] = None
+        return Step.save_model(self, model, *args, **kwargs)
+
+    def save_bkg(self, bkg_list, suffix="badpix_selfcal_bkg"):
+        """Save the background exposures to file with correct indexing
+        """
+        for i, bkg_model in enumerate(bkg_list):
+            self.save_model(bkg_model, suffix=suffix + f"_{str(i)}")
+
+    def process(self, input, selfcal_list=None, bkg_list=None):
+        """
+        Flag residual artifacts as bad pixels in the DQ array of a JWST exposure
+
+        Parameters
+        ----------
+        input: JWST data model or association
+            input science data to be corrected, or tuple of (sci, bkg, selfcal)
+
+        selfcal_list: list of ImageModels or filenames, default None
+            exposures to include as part of median background model used to find bad pixels,
+            but that are not flagged and returned as background exposures
+
+        bkg_list: list of ImageModels or filenames, default None
+            exposures to include as part of median background model used to find bad pixels,
+            and that are flagged and returned as background exposures
+
+        Returns
+        -------
+        output: JWST data model or association
+            data model with CRs flagged
+
+        Notes
+        -----
+        If an association file is read in, all exposures in the
+        association file, including science, background, and selfcal exposures,
+        are included in the MIN frame from which outliers are detected.
+        If selfcal_list and/or bkg_list are specified manually, they are appended to any
+        selfcal or background exposures found in the input association file.
+        If selfcal_list and bkg_list are both set to None and input is
+        a single science exposure, the step will be skipped with a warning unless
+        the force_single parameter is set True.
+        In that case, the input exposure will be used as the sole background exposure,
+        i.e., true self-calibration.
+        """
+        input_sci, selfcal_list, bkg_list = _parse_inputs(input, selfcal_list, bkg_list)
+
+        # ensure that there are background exposures to use, otherwise skip the step
+        # unless forced
+        if (len(selfcal_list + bkg_list) == 0) and (not self.force_single):
+            self.log.warning("No selfcal or background exposures provided for self-calibration. "
+                        "Skipping step. If you want to force self-calibration with the science "
+                        "exposure alone (generally not recommended), set force_single=True.")
```

Comment:
This should probably be broken up into 2 or 3 separate message lines, rather than one big long one.